tm2py/config.py

```diff
@@ -0,0 +1,550 @@
+
+    Args:
+        kwargs: input dictionary loaded from one or more TOML files
+    """
+
+    def __init__(self, kwargs: dict):
+        def get_type(otype):
+            try:
+                return otype.__origin__
+            except AttributeError:
+                return otype
+
+        fields = _get_fields(self)
+        extra_keys = set(kwargs.keys()) - set(f.name for f in fields)
+        if extra_keys:
+            raise Exception(f"unexpected key(s): {', '.join(extra_keys)}")
+        for field in fields:
+            value = kwargs.get(field.name, _MISSING)
+            if value is _MISSING:
+                if field.default is _MISSING:
+                    raise Exception(f"missing key {field.name}")
+                self.__dict__[field.name] = field.default
+                continue
+            try:
+                atype = get_type(field.type)
+                value = atype(value)
+                if isinstance(value, tuple):
+                    item_atype = get_type(field.type.__args__[0])
+                    processed_items = []
+                    for i, item in enumerate(value):
+                        try:
+                            processed_items.append(item_atype(item))
+                        except Exception as error:
+                            raise Exception(f"[{i}]: {error}") from error
+                    value = tuple(processed_items)
+                self.__dict__[field.name] = value
+            except Exception as error:
+                raise Exception(f"{field.name}: {error}") from error
+        self._validate()
+
+    def _validate(self):
+        pass
+
+    def __getitem__(self, key):
+        return getattr(self, key)
+
+    def items(self):
+        """D.items() -> a set-like object providing a view on D's items"""
+        return self.__dict__.items()
+
+    def get(self, key, default=None):
+        """Return the value for key if key is in the dictionary, else default."""
+        return self.__dict__.get(key, default)
+
+
+@dataclass(init=False, frozen=True)
+class Scenario(ConfigItem):
+    """Scenario parameters"""
+
+    year: int
+    verify: bool
+    maz_landuse_file: str
+
+
+@dataclass(init=False, frozen=True)
+class Run(ConfigItem):
+    """Model run parameters"""
+
+    start_iteration: int
+    end_iteration: int
+    start_component: str
+    initial_components: Tuple[str]
+    global_iteration_components: Tuple[str]
+    final_components: Tuple[str]
+
+
+@dataclass(init=False, frozen=True)
+class TimePeriod(ConfigItem):
+    """Time _time_period entry"""
+
+    name: str
+    length_hours: float
+    highway_capacity_factor: float
+    emme_scenario_id: int
+
+
+@dataclass(init=False, frozen=True)
+class Household(ConfigItem):
+    """Household (residents) model parameters"""
+
+    highway_demand_file: str
+    transit_demand_file: str
+
+
+@dataclass(init=False, frozen=True)
+class AirPassengerDemandAggregation(ConfigItem):
```

Comment:
For legibility outside this module (i.e. if class is imported and used elsewhere), I think these need need to have names which have Config in them even though it is explicit *here* that it is a `ConfigItem`

I came across this when I was doing some auto-documentation of classes in mermaid and was confused :-)

Reply:
cc: @lmz to see if she agrees

Reply:
Agree!